# Interview Case Study - Requirements

### 01.Data Loading: Load the provided three files using either SQL or PySpark. Name each table with a raw_ prefix to differentiate between original and transformed data.

### 02.Data Review and Storage: Review the loaded data and assign appropriate data types based on your best judgment. Identify primary and foreign keys for each table. Store the transformed data using a store_ prefix.

### 03.Product Master Transformations: Perform the following transformations on the product master data and write the results into a table named publish_product: Replace NULL values in the Color field with N/A.

##### Enhance the ProductCategoryName field when it is NULL using the following logic:

##### If ProductSubCategoryName is in (‘Gloves’, ‘Shorts’, ‘Socks’, ‘Tights’, ‘Vests’), set ProductCategoryName to ‘Clothing’.

##### If ProductSubCategoryName is in (‘Locks’, ‘Lights’, ‘Headsets’, ‘Helmets’, ‘Pedals’, ‘Pumps’), set ProductCategoryName to ‘Accessories’.

##### If ProductSubCategoryName contains the word ‘Frames’ or is in (‘Wheels’, ‘Saddles’), set ProductCategoryName to ‘Components’.

### 04.Sales Order Transformations: 
##### Join SalesOrderDetail with SalesOrderHeader on SalesOrderId and apply the following transformations: Calculate LeadTimeInBusinessDays as the difference between OrderDate and ShipDate, excluding Saturdays and Sundays 
##### Calculate TotalLineExtendedPrice using the formula: OrderQty * (UnitPrice - UnitPriceDiscount).
##### Write the results into a table named publish_orders, including: All fields from SalesOrderDetail. All fields from SalesOrderHeader except SalesOrderId, and rename Freight to TotalOrderFreight.

### 05.Analysis Questions: Provide answers to the following questions based on the transformed data:
##### Which color generated the highest revenue each year? 
##### What is the average LeadTimeInBusinessDays by ProductCategoryName?

# -> 01.Data Loading:
##### 01.Load the provided three files using either SQL or PySpark.
##### 02.Name each table with a raw_ prefix to differentiate between original and transformed data.

In [0]:
%sql
DROP TABLE prod_data.raw.raw_products;
DROP TABLE prod_data.raw.raw_sales_order_detail;
DROP TABLE prod_data.raw.raw_sales_order_header;
DROP TABLE prod_data.store.store_products;
DROP TABLE prod_data.store.store_sales_order_header;
DROP TABLE prod_data.store.store_sales_order_detail;
DROP TABLE prod_data.publish.publish_products;
DROP TABLE prod_data.publish.publish_orders;

In [0]:
%sql
CREATE CATALOG IF NOT EXISTS prod_data;
CREATE SCHEMA IF NOT EXISTS prod_data.raw;
CREATE VOLUME IF NOT EXISTS prod_data.raw.data;

In [0]:
%python

prod_data_path = "/Volumes/prod_data/raw/data/"

def read_csv_as_string(path):
    return (
        spark.read.format("csv")                 
            .option("header", True)                
            .option("inferSchema", False)           # Lets get everything as STRING so I can check each column later.
            .option("encoding", "utf-8")            # Ensures correct reading of international characters
            .load(path)                             # Load the file
    )

products_df = read_csv_as_string(prod_data_path + "products.csv")
sales_order_header_df = read_csv_as_string(prod_data_path + "sales_order_header.csv")
sales_order_detail_df = read_csv_as_string(prod_data_path + "sales_order_detail.csv")

In [0]:
display(products_df.limit(100))
display(sales_order_header_df.limit(100))
display(sales_order_detail_df.limit(100))

ProductID ProductDesc ProductNumber MakeFlag Color SafetyStockLevel ReorderPoint StandardCost ListPrice Size SizeUnitMeasureCode Weight WeightUnitMeasureCode ProductCategoryName ProductSubCategoryName 680 HL Road Frame - Black, 58 FR-R92B-58 True Black 500 375 1059.31 1431.5 58 CM 2.24 LB null Road Frames 706 HL Road Frame - Red, 58 FR-R92R-58 True Red 500 375 1059.31 1431.5 58 CM 2.24 LB null Road Frames 707 Sport-100 Helmet, Red HL-U509-R False Red 4 3 13.0863 34.99 null null null null null Helmets 708 Sport-100 Helmet, Black HL-U509 False Black 4 3 13.0863 34.99 null null null null null Helmets 709 Mountain Bike Socks, M SO-B909-M False White 4 3 3.3963 9.5 M null null null null Socks 710 Mountain Bike Socks, L SO-B909-L False White 4 3 3.3963 9.5 L null null null null Socks 711 Sport-100 Helmet, Blue HL-U509-B False Blue 4 3 13.0863 34.99 null null null null null Helmets 712 AWC Logo Cap CA-1098 False Multi 4 3 6.9223 8.99 null null null null null Caps 713 Long-Sleeve Logo Jersey, S LJ-0192-S False Multi 4 3 38.4923 49.99 S null null null null Jerseys 714 Long-Sleeve Logo Jersey, M LJ-0192-M False Multi 4 3 38.4923 49.99 M null null null null Jerseys 715 Long-Sleeve Logo Jersey, L LJ-0192-L False Multi 4 3 38.4923 49.99 L null null null null Jerseys 716 Long-Sleeve Logo Jersey, XL LJ-0192-X False Multi 4 3 38.4923 49.99 XL null null null null Jerseys 717 HL Road Frame - Red, 62 FR-R92R-62 True Red 500 375 868.6342 1431.5 62 CM 2.30 LB null Road Frames 718 HL Road Frame - Red, 44 FR-R92R-44 True Red 500 375 868.6342 1431.5 44 CM 2.12 LB null Road Frames 719 HL Road Frame - Red, 48 FR-R92R-48 True Red 500 375 868.6342 1431.5 48 CM 2.16 LB null Road Frames 720 HL Road Frame - Red, 52 FR-R92R-52 True Red 500 375 868.6342 1431.5 52 CM 2.20 LB null Road Frames 721 HL Road Frame - Red, 56 FR-R92R-56 True Red 500 375 868.6342 1431.5 56 CM 2.24 LB null Road Frames 722 LL Road Frame - Black, 58 FR-R38B-58 True Black 500 375 204.6251 337.22 58 CM 2.46 LB null Road Frames 723 LL Road Frame - Black, 60 FR-R38B-60 True Black 500 375 204.6251 337.22 60 CM 2.48 LB null Road Frames 724 LL Road Frame - Black, 62 FR-R38B-62 True Black 500 375 204.6251 337.22 62 CM 2.50 LB null Road Frames 725 LL Road Frame - Red, 44 FR-R38R-44 True Red 500 375 187.1571 337.22 44 CM 2.32 LB null Road Frames 726 LL Road Frame - Red, 48 FR-R38R-48 True Red 500 375 187.1571 337.22 48 CM 2.36 LB null Road Frames 727 LL Road Frame - Red, 52 FR-R38R-52 True Red 500 375 187.1571 337.22 52 CM 2.40 LB null Road Frames 728 LL Road Frame - Red, 58 FR-R38R-58 True Red 500 375 187.1571 337.22 58 CM 2.46 LB null Road Frames 729 LL Road Frame - Red, 60 FR-R38R-60 True Red 500 375 187.1571 337.22 60 CM 2.48 LB null Road Frames 730 LL Road Frame - Red, 62 FR-R38R-62 True Red 500 375 187.1571 337.22 62 CM 2.50 LB null Road Frames 731 ML Road Frame - Red, 44 FR-R72R-44 True Red 500 375 352.1394 594.83 44 CM 2.22 LB null Road Frames 732 ML Road Frame - Red, 48 FR-R72R-48 True Red 500 375 352.1394 594.83 48 CM 2.26 LB null Road Frames 733 ML Road Frame - Red, 52 FR-R72R-52 True Red 500 375 352.1394 594.83 52 CM 2.30 LB null Road Frames 734 ML Road Frame - Red, 58 FR-R72R-58 True Red 500 375 352.1394 594.83 58 CM 2.36 LB null Road Frames 735 ML Road Frame - Red, 60 FR-R72R-60 True Red 500 375 352.1394 594.83 60 CM 2.38 LB null Road Frames 736 LL Road Frame - Black, 44 FR-R38B-44 True Black 500 375 204.6251 337.22 44 CM 2.32 LB null Road Frames 737 LL Road Frame - Black, 48 FR-R38B-48 True Black 500 375 204.6251 337.22 48 CM 2.36 LB null Road Frames 738 LL Road Frame - Black, 52 FR-R38B-52 True Black 500 375 204.6251 337.22 52 CM 2.40 LB null Road Frames 739 HL Mountain Frame - Silver, 42 FR-M94S-42 True Silver 500 375 747.2002 1364.5 42 CM 2.72 LB null Mountain Frames 740 HL Mountain Frame - Silver, 44 FR-M94S-44 True Silver 500 375 706.811 1364.5 44 CM 2.76 LB null Mountain Frames 741 HL Mountain Frame - Silver, 48 FR-M94S-52 True Silver 500 375 706.811 1364.5 48 CM 2.80 LB null 

SalesOrderID OrderDate ShipDate OnlineOrderFlag AccountNumber CustomerID SalesPersonID Freight 43828 2021-06 2021-07-05 True 10-4030-027605 27605 null 89.4568 43829 2021-06 2021-07-05 True 10-4030-027611 27611 null 89.4568 43830 2021-06 2021-07-05 True 10-4030-016347 16347 null 89.4568 43831 2021-06 2021-07-05 True 10-4030-011028 11028 null 84.3748 43832 2021-06 2021-07-05 True 10-4030-013584 13584 null 89.4568 43659 2021-05-31 2021-06-07 False 10-4020-000676 29825 279 616.0984 43660 2021-05-31 2021-06-07 False 10-4020-000117 29672 279 38.8276 43661 2021-05-31 2021-06-07 False 10-4020-000442 29734 282 985.553 43662 2021-05-31 2021-06-07 False 10-4020-000227 29994 282 867.2389 43663 2021-05-31 2021-06-07 False 10-4020-000510 29565 276 12.5838 43664 2021-05-31 2021-06-07 False 10-4020-000397 29898 280 732.81 43665 2021-05-31 2021-06-07 False 10-4020-000146 29580 283 429.9821 43666 2021-05-31 2021-06-07 False 10-4020-000511 30052 276 151.9921 43667 2021-05-31 2021-06-07 False 10-4020-000646 29974 277 183.1626 43668 2021-05-31 2021-06-07 False 10-4020-000514 29614 282 1081.8017 43669 2021-05-31 2021-06-07 False 10-4020-000578 29747 283 22.0367 43670 2021-05-31 2021-06-07 False 10-4020-000504 29566 275 183.6126 43671 2021-05-31 2021-06-07 False 10-4020-000200 29890 283 244.0042 43672 2021-05-31 2021-06-07 False 10-4020-000119 30067 282 183.6257 43673 2021-05-31 2021-06-07 False 10-4020-000618 29844 275 111.8629 43674 2021-05-31 2021-06-07 False 10-4020-000083 29596 282 78.7315 43675 2021-05-31 2021-06-07 False 10-4020-000670 29827 277 170.8987 43676 2021-05-31 2021-06-07 False 10-4020-000017 29811 275 426.0206 43677 2021-05-31 2021-06-07 False 10-4020-000679 29824 278 233.4691 43678 2021-05-31 2021-06-07 False 10-4020-000203 29889 281 294.3981 43679 2021-05-31 2021-06-07 False 10-4020-000480 29761 278 39.3135 43680 2021-05-31 2021-06-07 False 10-4020-000491 29489 281 341.7623 43681 2021-05-31 2021-06-07 False 10-4020-000423 29661 279 413.4584 43682 2021-05-31 2021-06-07 False 10-4020-000486 29759 275 116.1876 43683 2021-05-31 2021-06-07 False 10-4020-000506 29497 283 1283.484 43684 2021-05-31 2021-06-07 False 10-4020-000549 29912 279 167.8941 43685 2021-05-31 2021-06-07 False 10-4020-000458 30084 279 82.2503 43686 2021-05-31 2021-06-07 False 10-4020-000344 29606 283 103.899 43687 2021-05-31 2021-06-07 False 10-4020-000269 29610 275 37.7513 43688 2021-05-31 2021-06-07 False 10-4020-000161 29718 275 382.1512 43689 2021-05-31 2021-06-07 False 10-4020-000166 29646 277 1022.3419 43690 2021-05-31 2021-06-07 False 10-4020-000431 29533 275 38.8133 43691 2021-05-31 2021-06-07 False 10-4020-000292 29570 277 26.2438 43692 2021-05-31 2021-06-07 False 10-4020-000221 29992 281 1155.1244 43693 2021-05-31 2021-06-07 False 10-4020-000485 29491 275 615.4864 43694 2021-05-31 2021-06-07 False 10-4020-000315 29549 279 618.2268 43695 2021-05-31 2021-06-07 False 10-4020-000027 29958 279 1183.5823 43696 2021-05-31 2021-06-07 False 10-4020-000603 29849 279 12.5838 43697 2021-05-31 2021-06-07 True 10-4030-021768 21768 null 89.4568 43698 2021-05-31 2021-06-07 True 10-4030-028389 28389 null 84.9998 43699 2021-05-31 2021-06-07 True 10-4030-025863 25863 null 84.9998 43700 2021-05-31 2021-06-07 True 10-4030-014501 14501 null 17.4775 43701 2021-05-31 2021-06-07 True 10-4030-011003 11003 null 84.9998 43702 2021-06-01 2021-06-08 True 10-4030-027645 27645 null 89.4568 43703 2021-06-01 2021-06-08 True 10-4030-016624 16624 null 89.4568 43704 2021-06-01 2021-06-08 True 10-4030-011005 11005 null 84.3748 43705 2021-06-01 2021-06-08 True 10-4030-011011 11011 null 84.9998 43706 2021-06-02 2021-06-09 True 10-4030-027621 27621 null 89.4568 43707 2021-06-02 2021-06-09 True 10-4030-027616 27616 null 89.4568 43708 2021-06-02 2021-06-09 True 10-4030-020042 20042 null 17.4775 43709 2021-06-02 2021-06-09 True 10-4030-016351 16351 null 89.4568 43710 2021-06-02 2021-06-09 True 10-4030-016517 16517 null 89.4568 43711 2021-06-03 2021-06-10 True 10-4030-027606 27606 null 89.4568 43712 

SalesOrderID SalesOrderDetailID OrderQty ProductID UnitPrice UnitPriceDiscount 43659 1 1 776 2024.9940 .0000 43659 2 3 777 2024.9940 .0000 43659 3 1 778 2024.9940 .0000 43659 4 1 771 2039.9940 .0000 43659 5 1 772 2039.9940 .0000 43659 6 2 773 2039.9940 .0000 43659 7 1 774 2039.9940 .0000 43659 8 3 714 28.8404 .0000 43659 9 1 716 28.8404 .0000 43659 10 6 709 5.7000 .0000 43659 11 2 712 5.1865 .0000 43659 12 4 711 20.1865 .0000 43660 13 1 762 419.4589 .0000 43660 14 1 758 874.7940 .0000 43661 15 1 745 809.7600 .0000 43661 16 1 743 714.7043 .0000 43661 17 2 747 714.7043 .0000 43661 18 4 712 5.1865 .0000 43661 19 4 715 28.8404 .0000 43661 20 2 742 722.5949 .0000 43661 21 3 775 2024.9940 .0000 43661 22 2 778 2024.9940 .0000 43661 23 2 711 20.1865 .0000 43661 24 2 741 .0000 .0000 43661 25 4 776 2024.9940 .0000 43661 26 2 773 2039.9940 .0000 43661 27 2 716 28.8404 .0000 43661 28 2 777 2024.9940 .0000 43661 29 5 708 20.1865 .0000 43662 30 3 764 419.4589 .0000 43662 31 5 770 419.4589 .0000 43662 32 2 730 183.9382 .0000 43662 33 4 754 874.7940 .0000 43662 34 3 725 183.9382 .0000 43662 35 5 762 419.4589 .0000 43662 36 3 765 419.4589 .0000 43662 37 2 768 419.4589 .0000 43662 38 1 753 2146.9620 .0000 43662 39 1 756 874.7940 .0000 43662 40 3 763 419.4589 .0000 43662 41 1 732 356.8980 .0000 43662 42 6 758 874.7940 .0000 43662 43 1 729 183.9382 .0000 43662 44 3 722 178.5808 .0000 43662 45 1 749 2146.9620 .0000 43662 46 3 760 419.4589 .0000 43662 47 1 726 183.9382 .0000 43662 48 1 733 356.8980 .0000 43662 49 1 738 178.5808 .0000 43662 50 3 766 419.4589 .0000 43662 51 1 755 874.7940 .0000 43663 52 1 760 419.4589 .0000 43664 53 1 772 2039.9940 .0000 43664 54 4 775 2024.9940 .0000 43664 55 1 714 28.8404 .0000 43664 56 1 716 28.8404 .0000 43664 57 2 777 2024.9940 .0000 43664 58 3 771 2039.9940 .0000 43664 59 1 773 2039.9940 .0000 43664 60 1 778 2024.9940 .0000 43665 61 2 711 20.1865 .0000 43665 62 1 773 2039.9940 .0000 43665 63 1 707 20.1865 .0000 43665 64 2 715 28.8404 .0000 43665 65 2 777 2024.9940 .0000 43665 66 2 712 5.1865 .0000 43665 67 2 775 2024.9940 .0000 43665 68 1 778 2024.9940 .0000 43665 69 6 709 5.7000 .0000 43665 70 1 776 2024.9940 .0000 43666 71 1 764 419.4589 .0000 43666 72 1 753 2146.9620 .0000 43666 73 1 732 356.8980 .0000 43666 74 1 756 874.7940 .0000 43666 75 2 768 419.4589 .0000 43666 76 1 766 419.4589 .0000 43667 77 3 710 5.7000 .0000 43667 78 1 773 2039.9940 .0000 43667 79 1 778 2024.9940 .0000 43667 80 1 775 2024.9940 .0000 43668 81 3 756 874.7940 .0000 43668 82 2 753 2146.9620 .0000 43668 83 7 760 419.4589 .0000 43668 84 6 765 419.4589 .0000 43668 85 6 715 28.8404 .0000 43668 86 6 730 183.9382 .0000 43668 87 2 707 20.1865 .0000 43668 88 2 711 20.1865 .0000 43668 89 2 754 874.7940 .0000 43668 90 4 712 5.1865 .0000 43668 91 2 729 183.9382 .0000 43668 92 3 755 874.7940 .0000 43668 93 3 761 419.4589 .0000 43668 94 2 770 419.4589 .0000 43668 95 3 726 183.9382 .0000 43668 96 2 764 419.4589 .0000 43668 97 2 766 419.4589 .0000 43668 98 2 725 183.9382 .0000 43668 99 1 716 28.8404 .0000 43668 100 2 768 419.4589 .0000

In [0]:
%python
products_df.write.format("delta").mode("overwrite").saveAsTable("prod_data.raw.raw_products")
sales_order_header_df.write.format("delta").mode("overwrite").saveAsTable("prod_data.raw.raw_sales_order_header")
sales_order_detail_df.write.format("delta").mode("overwrite").saveAsTable("prod_data.raw.raw_sales_order_detail")

# -> 02.Data Review and Storage:
##### 01.Review the loaded data and assign appropriate data types based on your best judgment.
##### 02.Identify primary and foreign keys for each table.
##### 03.Store the transformed data using a store_ prefix.

## -> 02.1 Defining data types for each table.
### Defining DATA types for each one of the tables and separating it into groups for TEST CASES.

In [0]:
# Int Columns
int_cols = ["ProductID", "SafetyStockLevel", "ReorderPoint"] 

# Decimal Columns
decimal_cols = ["StandardCost", "ListPrice", "Weight"]  # Defined as DECIMAL since we want to keep the precision of those columns tracked down.

# BOOLEAN columns
bool_cols = ["MakeFlag"]

# String columns
string_cols = [
    "ProductDesc",
    "ProductNumber",
    "Color",
    "Size",
    "SizeUnitMeasureCode",
    "WeightUnitMeasureCode",
    "ProductCategoryName",
    "ProductSubCategoryName",
]

In [0]:
# Int Columns
int_cols_header = [
    "SalesOrderID",
    "CustomerID",
    "SalesPersonID"
]

# Decimal Columns
decimal_cols_header = [
    "Freight"
]

# Boolean Columns
bool_cols_header = [
    "OnlineOrderFlag"
]

# String Columns
string_cols_header = [
    "AccountNumber"
]

# Date Columns
date_cols_header = [
    "ShipDate"
]

# Timestamp Columns
timestamp_cols_header = [
    "OrderDate"
]

In [0]:
# Int Columns
int_cols_detail = [
    "SalesOrderDetailID",
    "SalesOrderID",
    "OrderQty",
    "ProductID"
]

# Decimal Columns 
decimal_cols_detail_normal = ["UnitPrice"]

# Decimal Columns for UnityPriceDiscount which has a different format since it's a fraction representation.
decimal_cols_detail_unipd = ["UnitPriceDiscount"]

## -> 02.2 Test Cases
### Creating test cases to verify integrity and issues in each segment of group from the tables.

In [0]:
from pyspark.sql import functions as F

'''
#1 Test case to validate if ProductID has duplicates
'''
def duplicate_productid_condition(df, col="ProductID"):
    dup_ids = (df.groupBy(col).count().filter("count > 1").select(col)) 

    # Join to mark the duplicate rows
    return F.col(col).isin([row[col] for row in dup_ids.collect()])

'''
#2 Test case to validate if there is any INT column with issue:
'''
def invalid_int_condition(col):
    return (F.col(col).isNotNull() &
            ~F.col(col).rlike(r"^[0-9]+$"))
    
#3 Test case to validate if there is any DECIMAL column with issue:
def invalid_decimal_condition(col):
    return (F.col(col).isNotNull() &
            ~F.col(col).rlike(r"^[0-9]+([.,][0-9]+)?$"))

# 3.1 Test case to validate the column UNITYPRICEDESCOUNT because it looks like a % at first but reviewing it I beliave it's the actually value of discount.
def invalid_decimal_condition_unipd(col):
    valid_decimal_pattern = r"^[0-9]*([.,][0-9]+)?$"
    return (
        F.col(col).isNotNull() &
        ~F.col(col).rlike(valid_decimal_pattern)
    )
    
#4 Test case to validate if there is any BOOLEAN column with issue:
def invalid_bool_condition(col):
    valid = ["true","false"]
    return (F.col(col).isNotNull() &
            ~F.lower(F.col(col)).isin(valid))
    
#5 Test case to validate if there is any STRING column with SPACES issues:
def needs_trim_condition(col):
    return (F.col(col).isNotNull() &
            (F.col(col) != F.trim(F.col(col))))
    
# 6 Test case to validate DATE columns looking if they are not null and if spark can parse them as DATE
def invalid_date_condition(col):
    parsed = F.to_date(F.col(col))
    return F.col(col).isNotNull() & parsed.isNull()


# 7 Test case to validate DATE columns looking if they are not null and if spark can parse them as TIMESTAMP
def invalid_timestamp_condition(col):
    col_ref = F.col(col)
    parsed_ts = F.to_timestamp(col_ref)
    incomplete_ts = col_ref.rlike(r"^\d{4}-\d{2}$")
    return col_ref.isNotNull() & (parsed_ts.isNull() | incomplete_ts)

In [0]:
from pyspark.sql import functions as f

'''
This function will summarize and display the amount of erros when veryfy a table in the RAW catalog.
'''
def total_errors_from_products(df):
    total_rows = df.count()

    def print_metric(label, count):
        pct = (count / total_rows) * 100 if total_rows else 0
        print(f"{label}: {count} rows ({pct:.2f}%)")

    # 1) DUPLICATES (PK)
    duplicate_keys = (
        df.groupBy("ProductID").count().filter("count > 1").count()
    )

    # 2) INT COLUMNS
    int_error = None
    for c in int_cols:
        cond = invalid_int_condition(c)
        int_error = cond if int_error is None else (int_error | cond)

    # 3) DECIMAL COLUMNS
    decimal_error = None
    for c in decimal_cols:
        cond = invalid_decimal_condition(c)
        decimal_error = cond if decimal_error is None else (decimal_error | cond)

    # 4) BOOLEAN COLUMNS
    bool_error = None
    for c in bool_cols:
        cond = invalid_bool_condition(c)
        bool_error = cond if bool_error is None else (bool_error | cond)

    # 5) STRING COLUMNS
    trim_error = None
    for c in string_cols:
        cond = needs_trim_condition(c)
        trim_error = cond if trim_error is None else (trim_error | cond)

    # 3) Get all erros at once so we don't full_scan the table for each testcase.

    summary = df.select(
        int_error.cast("int").alias("int_error_int"),
        decimal_error.cast("int").alias("decimal_error_int"),
        bool_error.cast("int").alias("bool_error_int"),
        trim_error.cast("int").alias("trim_error_int")
    ).agg(
        F.sum("int_error_int").alias("int_errors"),
        F.sum("decimal_error_int").alias("decimal_errors"),
        F.sum("bool_error_int").alias("bool_errors"),
        F.sum("trim_error_int").alias("trim_errors")
    ).collect()[0]

    # 4) Display each metric for each test_case
    
    print("Test_Cases_Validations: RAW_products")
    print(f"Total rows: {total_rows}\n")
    print_metric("DUPLICATE ProductID (PK keys)", duplicate_keys)
    print_metric("INT errors", summary["int_errors"] or 0)
    print_metric("DECIMAL errors", summary["decimal_errors"] or 0)
    print_metric("BOOLEAN errors", summary["bool_errors"] or 0)
    print_metric("STRING issues", summary["trim_errors"] or 0)
    display(df.filter(trim_error))

In [0]:
from pyspark.sql import functions as F

'''
This function will summarize and display the amount of errors
when verifying the RAW_sales_order_header table.
'''
def total_errors_from_header(df):

    total_rows = df.count()

    def print_metric(label, count):
        pct = (count / total_rows) * 100 if total_rows else 0 #% errors
        print(f"{label}: {count} rows ({pct:.2f}%)")

    # 1) DUPLICATES (PK)
    duplicate_keys = (
        df.groupBy("SalesOrderID").count().filter("count > 1").count()
    )

    # 2) INT Columns
    int_error = None
    for c in int_cols_header:
        cond = invalid_int_condition(c)
        int_error = cond if int_error is None else (int_error | cond)

    # 3) DECIMAL Columns
    decimal_error = None
    for c in decimal_cols_header:
        cond = invalid_decimal_condition(c)
        decimal_error = cond if decimal_error is None else (decimal_error | cond)
    
    # 4) BOOLEAN Columns
    bool_error = None
    for c in bool_cols_header:
        cond = invalid_bool_condition(c)
        bool_error = cond if bool_error is None else (bool_error | cond)

    # 5) STRING Columns
    trim_error = None
    for c in string_cols_header:
        cond = needs_trim_condition(c)
        trim_error = cond if trim_error is None else (trim_error | cond)

    # 6) DATE Columns
    date_error = None
    for c in date_cols_header:
        cond = invalid_date_condition(c)
        date_error = cond if date_error is None else (date_error | cond)

    # 7) TIMESTAMP Columns
    timestamp_error = None
    for c in timestamp_cols_header:
        cond = invalid_timestamp_condition(c)
        timestamp_error = cond if timestamp_error is None else (timestamp_error | cond)

    # 8) Get all errors at once so we don't full_scan the table for each testcase.
    summary = df.select(
        int_error.cast("int").alias("int_error_int"),
        decimal_error.cast("int").alias("decimal_error_int"),
        bool_error.cast("int").alias("bool_error_int"),
        trim_error.cast("int").alias("trim_error_int"),
        date_error.cast("int").alias("date_error_int"),
        timestamp_error.cast("int").alias("timestamp_error_int")
    ).agg(
        F.sum("int_error_int").alias("int_errors"),
        F.sum("decimal_error_int").alias("decimal_errors"),
        F.sum("bool_error_int").alias("bool_errors"),
        F.sum("trim_error_int").alias("trim_errors"),
        F.sum("date_error_int").alias("date_errors"),
        F.sum("timestamp_error_int").alias("timestamp_errors")
    ).collect()[0]

    # 9) Display each metric for each test_case
    print("Test_Cases_Validations: RAW_sales_order_header")
    print(f"Total rows: {total_rows}\n")

    print_metric("DUPLICATE SalesOrderID (PK keys)", duplicate_keys)
    print_metric("INT errors", summary["int_errors"] or 0)
    print_metric("DECIMAL errors", summary["decimal_errors"] or 0)
    print_metric("BOOLEAN errors", summary["bool_errors"] or 0)
    print_metric("STRING issues", summary["trim_errors"] or 0)
    print_metric("DATE errors", summary["date_errors"] or 0)
    print_metric("TIMESTAMP errors", summary["timestamp_errors"] or 0)
    display(df.filter(timestamp_error))


In [0]:
from pyspark.sql import functions as F

'''
This function will summarize and display the amount of errors
when verifying the RAW_sales_order_detail table.
'''
def total_errors_from_detail(df):

    total_rows = df.count()

    def print_metric(label, count):
        pct = (count / total_rows) * 100 if total_rows else 0
        print(f"{label}: {count} rows ({pct:.2f}%)")

    # 1) DUPLICATES (PK)
    duplicate_keys = (
        df.groupBy("SalesOrderDetailID").count().filter("count > 1").count()
    )

    # 2) INT Columns
    int_error = None
    for c in int_cols_detail:
        cond = invalid_int_condition(c)
        int_error = cond if int_error is None else (int_error | cond)

    # 3) DECIMAL Columns (normal decimals)
    decimal_error = None
    for c in decimal_cols_detail_normal:   # <-- apenas ajuste do grupo
        cond = invalid_decimal_condition(c)
        decimal_error = cond if decimal_error is None else (decimal_error | cond)

    # 3.1) DECIMAL Columns (UnitPriceDiscount special case)
    decimal_error_unipd = None
    for c in decimal_cols_detail_unipd:     # <-- seu novo test case aplicado aqui
        cond = invalid_decimal_condition_unipd(c)
        decimal_error_unipd = cond if decimal_error_unipd is None else (decimal_error_unipd | cond)

    # 3) Get all erros at once so we don't full_scan the table for each testcase.
    summary = df.select(
        int_error.cast("int").alias("int_error_int"),
        decimal_error.cast("int").alias("decimal_error_int"),
        decimal_error_unipd.cast("int").alias("decimal_error_unipd_int")
    ).agg(
        F.sum("int_error_int").alias("int_errors"),
        F.sum("decimal_error_int").alias("decimal_errors"),
        F.sum("decimal_error_unipd_int").alias("decimal_unipd_errors")
    ).collect()[0]

    # 4) Display each metric for each test_case
    print("Test_Cases_Validations: RAW_sales_order_detail")
    print(f"Total rows: {total_rows}\n")

    print_metric("DUPLICATE SalesOrderDetailID (PK keys)", duplicate_keys)
    print_metric("INT errors", summary["int_errors"] or 0)
    print_metric("DECIMAL errors (normal)", summary["decimal_errors"] or 0)
    print_metric("DECIMAL errors (UnitPriceDiscount)", summary["decimal_unipd_errors"] or 0)


In [0]:
total_errors_from_products(products_df)

Test_Cases_Validations: RAW_products
Total rows: 303

DUPLICATE ProductID (PK keys): 8 rows (2.64%)
INT errors: 0 rows (0.00%)
DECIMAL errors: 0 rows (0.00%)
BOOLEAN errors: 0 rows (0.00%)
STRING issues: 198 rows (65.35%)


ProductID ProductDesc ProductNumber MakeFlag Color SafetyStockLevel ReorderPoint StandardCost ListPrice Size SizeUnitMeasureCode Weight WeightUnitMeasureCode ProductCategoryName ProductSubCategoryName 680 HL Road Frame - Black, 58 FR-R92B-58 True Black 500 375 1059.31 1431.5 58 CM 2.24 LB null Road Frames 706 HL Road Frame - Red, 58 FR-R92R-58 True Red 500 375 1059.31 1431.5 58 CM 2.24 LB null Road Frames 717 HL Road Frame - Red, 62 FR-R92R-62 True Red 500 375 868.6342 1431.5 62 CM 2.30 LB null Road Frames 718 HL Road Frame - Red, 44 FR-R92R-44 True Red 500 375 868.6342 1431.5 44 CM 2.12 LB null Road Frames 719 HL Road Frame - Red, 48 FR-R92R-48 True Red 500 375 868.6342 1431.5 48 CM 2.16 LB null Road Frames 720 HL Road Frame - Red, 52 FR-R92R-52 True Red 500 375 868.6342 1431.5 52 CM 2.20 LB null Road Frames 721 HL Road Frame - Red, 56 FR-R92R-56 True Red 500 375 868.6342 1431.5 56 CM 2.24 LB null Road Frames 722 LL Road Frame - Black, 58 FR-R38B-58 True Black 500 375 204.6251 337.22 58 CM 2.46 LB null Road Frames 723 LL Road Frame - Black, 60 FR-R38B-60 True Black 500 375 204.6251 337.22 60 CM 2.48 LB null Road Frames 724 LL Road Frame - Black, 62 FR-R38B-62 True Black 500 375 204.6251 337.22 62 CM 2.50 LB null Road Frames 725 LL Road Frame - Red, 44 FR-R38R-44 True Red 500 375 187.1571 337.22 44 CM 2.32 LB null Road Frames 726 LL Road Frame - Red, 48 FR-R38R-48 True Red 500 375 187.1571 337.22 48 CM 2.36 LB null Road Frames 727 LL Road Frame - Red, 52 FR-R38R-52 True Red 500 375 187.1571 337.22 52 CM 2.40 LB null Road Frames 728 LL Road Frame - Red, 58 FR-R38R-58 True Red 500 375 187.1571 337.22 58 CM 2.46 LB null Road Frames 729 LL Road Frame - Red, 60 FR-R38R-60 True Red 500 375 187.1571 337.22 60 CM 2.48 LB null Road Frames 730 LL Road Frame - Red, 62 FR-R38R-62 True Red 500 375 187.1571 337.22 62 CM 2.50 LB null Road Frames 731 ML Road Frame - Red, 44 FR-R72R-44 True Red 500 375 352.1394 594.83 44 CM 2.22 LB null Road Frames 732 ML Road Frame - Red, 48 FR-R72R-48 True Red 500 375 352.1394 594.83 48 CM 2.26 LB null Road Frames 733 ML Road Frame - Red, 52 FR-R72R-52 True Red 500 375 352.1394 594.83 52 CM 2.30 LB null Road Frames 734 ML Road Frame - Red, 58 FR-R72R-58 True Red 500 375 352.1394 594.83 58 CM 2.36 LB null Road Frames 735 ML Road Frame - Red, 60 FR-R72R-60 True Red 500 375 352.1394 594.83 60 CM 2.38 LB null Road Frames 736 LL Road Frame - Black, 44 FR-R38B-44 True Black 500 375 204.6251 337.22 44 CM 2.32 LB null Road Frames 737 LL Road Frame - Black, 48 FR-R38B-48 True Black 500 375 204.6251 337.22 48 CM 2.36 LB null Road Frames 738 LL Road Frame - Black, 52 FR-R38B-52 True Black 500 375 204.6251 337.22 52 CM 2.40 LB null Road Frames 739 HL Mountain Frame - Silver, 42 FR-M94S-42 True Silver 500 375 747.2002 1364.5 42 CM 2.72 LB null Mountain Frames 740 HL Mountain Frame - Silver, 44 FR-M94S-44 True Silver 500 375 706.811 1364.5 44 CM 2.76 LB null Mountain Frames 741 HL Mountain Frame - Silver, 48 FR-M94S-52 True Silver 500 375 706.811 1364.5 48 CM 2.80 LB null Mountain Frames 742 HL Mountain Frame - Silver, 46 FR-M94S-46 True Silver 500 375 747.2002 1364.5 46 CM 2.84 LB null Mountain Frames 743 HL Mountain Frame - Black, 42 FR-M94B-42 True Black 500 375 739.041 1349.6 42 CM 2.72 LB null Mountain Frames 744 HL Mountain Frame - Black, 44 FR-M94B-44 True Black 500 375 699.0928 1349.6 44 CM 2.76 LB null Mountain Frames 745 HL Mountain Frame - Black, 48 FR-M94B-48 True Black 500 375 699.0928 1349.6 48 CM 2.80 LB null Mountain Frames 746 HL Mountain Frame - Black, 46 FR-M94B-46 True Black 500 375 739.041 1349.6 46 CM 2.84 LB null Mountain Frames 747 HL Mountain Frame - Black, 38 FR-M94B-38 True Black 500 375 739.041 1349.6 38 CM 2.68 LB null Mountain Frames 748 HL Mountain Frame - Silver, 38 FR-M94S-38 True Silver 500 375 747.2002 1364.5 38 CM 2.68 LB null Mountain Frames 749 Road-150 Red, 62 BK-R93R-62 True Red 100 75 2171.2942 3578.27 62 CM 15.00 LB Components Road Bikes 750 Road-150 Red, 44 BK-R93R-44 True Red 10

In [0]:
total_errors_from_header(sales_order_header_df)

Test_Cases_Validations: RAW_sales_order_header
Total rows: 31465

DUPLICATE SalesOrderID (PK keys): 0 rows (0.00%)
INT errors: 0 rows (0.00%)
DECIMAL errors: 0 rows (0.00%)
BOOLEAN errors: 0 rows (0.00%)
STRING issues: 0 rows (0.00%)
DATE errors: 0 rows (0.00%)
TIMESTAMP errors: 5 rows (0.02%)


SalesOrderID OrderDate ShipDate OnlineOrderFlag AccountNumber CustomerID SalesPersonID Freight 43828 2021-06 2021-07-05 True 10-4030-027605 27605 null 89.4568 43829 2021-06 2021-07-05 True 10-4030-027611 27611 null 89.4568 43830 2021-06 2021-07-05 True 10-4030-016347 16347 null 89.4568 43831 2021-06 2021-07-05 True 10-4030-011028 11028 null 84.3748 43832 2021-06 2021-07-05 True 10-4030-013584 13584 null 89.4568

In [0]:
total_errors_from_detail(sales_order_detail_df)

Test_Cases_Validations: RAW_sales_order_detail
Total rows: 121317

DUPLICATE SalesOrderDetailID (PK keys): 0 rows (0.00%)
INT errors: 2 rows (0.00%)
DECIMAL errors (normal): 94 rows (0.08%)
DECIMAL errors (UnitPriceDiscount): 0 rows (0.00%)


In [0]:
%sql
SELECT 
  ProductID, 
  COUNT(*) AS ProductID_Duplicates 
FROM 
  prod_data.raw.raw_products 
GROUP BY 
  ProductID
HAVING 
  COUNT(*) > 1

-- SELECT *
-- FROM prod_data.raw.raw_products
-- WHERE 
--       ProductDesc IS NULL
--    OR ProductNumber IS NULL
--    OR StandardCost IS NULL
--    OR ListPrice IS NULL
--    OR Color IS NULL
--    OR ProductCategoryName IS NULL
--    OR ProductSubCategoryName IS NULL;
--LIMIT 100

ProductID ProductID_Duplicates 713 2 714 2 715 2 716 2 881 2 882 2 883 2 884 2

In [0]:
%sql
SELECT
  *
FROM 
  prod_data.raw.raw_products
WHERE 
  ProductID = 713

ProductID ProductDesc ProductNumber MakeFlag Color SafetyStockLevel ReorderPoint StandardCost ListPrice Size SizeUnitMeasureCode Weight WeightUnitMeasureCode ProductCategoryName ProductSubCategoryName 713 Long-Sleeve Logo Jersey, S LJ-0192-S False Multi 4 3 38.4923 49.99 S null null null null Jerseys 713 Long-Sleeve Logo Jersey, S LJ-0192-S False Multi 4 3 38.4923 49.99 S null null null Clothing Shirt

In [0]:
%sql
SELECT *
FROM prod_data.raw.raw_products
WHERE 1=1 
 --OR ProductDesc <> TRIM(ProductDesc)
 --OR ProductCategoryName <> TRIM(ProductCategoryName)
 --OR ProductSubCategoryName <> TRIM(ProductSubCategoryName)
 OR WeightUnitMeasureCode <> TRIM(WeightUnitMeasureCode)
 OR SizeUnitMeasureCode <> TRIM(SizeUnitMeasureCode)
LIMIT 100;

ProductID ProductDesc ProductNumber MakeFlag Color SafetyStockLevel ReorderPoint StandardCost ListPrice Size SizeUnitMeasureCode Weight WeightUnitMeasureCode ProductCategoryName ProductSubCategoryName 680 HL Road Frame - Black, 58 FR-R92B-58 True Black 500 375 1059.31 1431.5 58 CM 2.24 LB null Road Frames 706 HL Road Frame - Red, 58 FR-R92R-58 True Red 500 375 1059.31 1431.5 58 CM 2.24 LB null Road Frames 707 Sport-100 Helmet, Red HL-U509-R False Red 4 3 13.0863 34.99 null null null null null Helmets 708 Sport-100 Helmet, Black HL-U509 False Black 4 3 13.0863 34.99 null null null null null Helmets 709 Mountain Bike Socks, M SO-B909-M False White 4 3 3.3963 9.5 M null null null null Socks 710 Mountain Bike Socks, L SO-B909-L False White 4 3 3.3963 9.5 L null null null null Socks 711 Sport-100 Helmet, Blue HL-U509-B False Blue 4 3 13.0863 34.99 null null null null null Helmets 712 AWC Logo Cap CA-1098 False Multi 4 3 6.9223 8.99 null null null null null Caps 713 Long-Sleeve Logo Jersey, S LJ-0192-S False Multi 4 3 38.4923 49.99 S null null null null Jerseys 714 Long-Sleeve Logo Jersey, M LJ-0192-M False Multi 4 3 38.4923 49.99 M null null null null Jerseys 715 Long-Sleeve Logo Jersey, L LJ-0192-L False Multi 4 3 38.4923 49.99 L null null null null Jerseys 716 Long-Sleeve Logo Jersey, XL LJ-0192-X False Multi 4 3 38.4923 49.99 XL null null null null Jerseys 717 HL Road Frame - Red, 62 FR-R92R-62 True Red 500 375 868.6342 1431.5 62 CM 2.30 LB null Road Frames 718 HL Road Frame - Red, 44 FR-R92R-44 True Red 500 375 868.6342 1431.5 44 CM 2.12 LB null Road Frames 719 HL Road Frame - Red, 48 FR-R92R-48 True Red 500 375 868.6342 1431.5 48 CM 2.16 LB null Road Frames 720 HL Road Frame - Red, 52 FR-R92R-52 True Red 500 375 868.6342 1431.5 52 CM 2.20 LB null Road Frames 721 HL Road Frame - Red, 56 FR-R92R-56 True Red 500 375 868.6342 1431.5 56 CM 2.24 LB null Road Frames 722 LL Road Frame - Black, 58 FR-R38B-58 True Black 500 375 204.6251 337.22 58 CM 2.46 LB null Road Frames 723 LL Road Frame - Black, 60 FR-R38B-60 True Black 500 375 204.6251 337.22 60 CM 2.48 LB null Road Frames 724 LL Road Frame - Black, 62 FR-R38B-62 True Black 500 375 204.6251 337.22 62 CM 2.50 LB null Road Frames 725 LL Road Frame - Red, 44 FR-R38R-44 True Red 500 375 187.1571 337.22 44 CM 2.32 LB null Road Frames 726 LL Road Frame - Red, 48 FR-R38R-48 True Red 500 375 187.1571 337.22 48 CM 2.36 LB null Road Frames 727 LL Road Frame - Red, 52 FR-R38R-52 True Red 500 375 187.1571 337.22 52 CM 2.40 LB null Road Frames 728 LL Road Frame - Red, 58 FR-R38R-58 True Red 500 375 187.1571 337.22 58 CM 2.46 LB null Road Frames 729 LL Road Frame - Red, 60 FR-R38R-60 True Red 500 375 187.1571 337.22 60 CM 2.48 LB null Road Frames 730 LL Road Frame - Red, 62 FR-R38R-62 True Red 500 375 187.1571 337.22 62 CM 2.50 LB null Road Frames 731 ML Road Frame - Red, 44 FR-R72R-44 True Red 500 375 352.1394 594.83 44 CM 2.22 LB null Road Frames 732 ML Road Frame - Red, 48 FR-R72R-48 True Red 500 375 352.1394 594.83 48 CM 2.26 LB null Road Frames 733 ML Road Frame - Red, 52 FR-R72R-52 True Red 500 375 352.1394 594.83 52 CM 2.30 LB null Road Frames 734 ML Road Frame - Red, 58 FR-R72R-58 True Red 500 375 352.1394 594.83 58 CM 2.36 LB null Road Frames 735 ML Road Frame - Red, 60 FR-R72R-60 True Red 500 375 352.1394 594.83 60 CM 2.38 LB null Road Frames 736 LL Road Frame - Black, 44 FR-R38B-44 True Black 500 375 204.6251 337.22 44 CM 2.32 LB null Road Frames 737 LL Road Frame - Black, 48 FR-R38B-48 True Black 500 375 204.6251 337.22 48 CM 2.36 LB null Road Frames 738 LL Road Frame - Black, 52 FR-R38B-52 True Black 500 375 204.6251 337.22 52 CM 2.40 LB null Road Frames 739 HL Mountain Frame - Silver, 42 FR-M94S-42 True Silver 500 375 747.2002 1364.5 42 CM 2.72 LB null Mountain Frames 740 HL Mountain Frame - Silver, 44 FR-M94S-44 True Silver 500 375 706.811 1364.5 44 CM 2.76 LB null Mountain Frames 741 HL Mountain Frame - Silver, 48 FR-M94S-52 True Silver 500 375 706.811 1364.5 48 CM 2.80 LB null 

### 2.11 Notes about data_validation
#### 01.raw_products
##### 1.1-8 duplicates on PRIMARY KEY for PRODUCTID with is 2.64% of the total table.
##### 1.2-No erros founded on INT, DECIMAL, BOOLEAN data fields.
##### 1.3-65.35% of STRING Columns with SPACES issues.
##### 1.3-Many columns with null values(must create another test case)
##### 1.4-Monetary values changed for DECIMAL
##### 1.5-Total of 295 distincts products.

#### 02.sales_order_header_df
##### 2.1- No Duplicates on PK
##### 2.2- No errors on INT,DECIMAL,BOOLEAN,STRING,DATA FIELDS
##### 2.3- There are 5 rows with ERROS on TIMESTAMP FIELD which represents 0.02% of the dataset. Those timestamp are returning as YYYY-MM with no valid format.
##### 2.4- A important point is that UnitPriceDiscount is actually represents a fraction of the discount, so the requirement is not correct when it says to
##### calculate like: TotalLineExtendedPrice using the formula: OrderQty * (UnitPrice - UnitPriceDiscount). This should probably be something like; OrderQty * UnitPrice * (1 - UnitPriceDiscount).

#### 03.sales_order_detail_df
##### 3.1 - There are 2 ERROS for INT COLUMNS
##### 3.2 - There are 94 errors/issues for DECIMAL COLUMNS which represents 0.08% of the total dataset. Those errors are different format within the decimal COLUMN
##### 3.3 - No other errors or attention points were found beyond that.

## ->02.3 Creating Store schema and tables
### Ingestion Raw -> Store

In [0]:
%sql
CREATE SCHEMA IF NOT EXISTS prod_data.store;

In [0]:
%sql
CREATE OR REPLACE TABLE prod_data.store.store_products (
  ProductID INT COMMENT "Primary key. Unique identifier for each product",
  ProductDesc STRING COMMENT "Normal description",
  ProductNumber STRING COMMENT "Alphanumeric column / SKU",
  MakeFlag BOOLEAN COMMENT "Internal flag",
  Color STRING COMMENT "Color of the product",
  SafetyStockLevel INT COMMENT "Minimum stock level required",
  ReorderPoint INT COMMENT "Reorder threshold",
  StandardCost DECIMAL(19,4) COMMENT "Internal product cost reference", --DECIMAL since is monetarial value
  ListPrice DECIMAL(19,4) COMMENT "Product price",                      -- DECIMAL since is monetarial value
  Size STRING COMMENT "Size label",
  SizeUnitMeasureCode STRING COMMENT "Unit of measure for Size",
  Weight DECIMAL(10,4) COMMENT "Product weight.",                       --TO DECIMAL and precision 4 since might be needed for any calculation.
  WeightUnitMeasureCode STRING COMMENT "Unit of measure for Weight",
  ProductCategoryName STRING COMMENT "Category name of a product",
  ProductSubCategoryName STRING COMMENT "SubCategory name of ap roduct", 
  CONSTRAINT PK_store_products PRIMARY KEY (ProductID)
)
USING DELTA;

In [0]:
%sql
--Ingestion from raw to store but also storing the data properly from RAW.
--REMOVING DUPLICATE ID's
--REMOVING UNECESSARY SPACES
INSERT INTO prod_data.store.store_products (
    ProductID,
    ProductDesc,
    ProductNumber,
    MakeFlag,
    Color,
    SafetyStockLevel,
    ReorderPoint,
    StandardCost,
    ListPrice,
    Size,
    SizeUnitMeasureCode,
    Weight,
    WeightUnitMeasureCode,
    ProductCategoryName,
    ProductSubCategoryName
)
WITH dedup_PRODUCTID AS (
    SELECT
        *,
        ROW_NUMBER() OVER (PARTITION BY ProductID 
        ORDER BY
                CASE WHEN ProductCategoryName IS NOT NULL THEN 0 ELSE 1 END,    -- Keep Category if category is not null
                CASE WHEN ProductSubCategoryName IS NOT NULL THEN 0 ELSE 1 END  -- Keep SubCategory if subcategory is not null
        ) AS choose_keep_line
    FROM prod_data.raw.raw_products
)
SELECT
    ProductID,
    ProductDesc AS ProductDesc,
    ProductNumber AS ProductNumber,
    MakeFlag,
    Color AS Color,
    SafetyStockLevel,
    ReorderPoint,
    StandardCost,
    ListPrice,
    Size AS Size,
    SizeUnitMeasureCode AS SizeUnitMeasureCode,
    Weight,
    WeightUnitMeasureCode AS WeightUnitMeasureCode,
    TRIM(ProductCategoryName) AS ProductCategoryName,
    TRIM(ProductSubCategoryName) AS ProductSubCategoryName
FROM 
    dedup_PRODUCTID
WHERE 
    choose_keep_line = 1;

num_affected_rows num_inserted_rows 295 295

In [0]:
%sql
CREATE OR REPLACE TABLE prod_data.store.store_sales_order_header (
  SalesOrderID INT COMMENT "Unique identifier for each sales order heade",
  OrderDate TIMESTAMP COMMENT "Creation time of an order",
  ShipDate DATE COMMENT "Date the order was shipped",
  OnlineOrderFlag BOOLEAN COMMENT "Identifier for online order or not",
  AccountNumber STRING COMMENT "Customer account",
  CustomerID INT COMMENT "Identifier of customer",
  SalesPersonID INT COMMENT "Identifier of SalesPerson",
  Freight DECIMAL(19,4) COMMENT "Freight cost associated with the order.",
  CONSTRAINT pk_store_sales_order_header PRIMARY KEY (SalesOrderID)
)
USING DELTA;

In [0]:
%sql
INSERT INTO prod_data.store.store_sales_order_header
SELECT 
  SalesOrderID,
  OrderDate,
  ShipDate,
  OnlineOrderFlag,
  AccountNumber,
  CustomerID,
  SalesPersonID,
  Freight
FROM 
  prod_data.raw.raw_sales_order_header;

num_affected_rows num_inserted_rows 31465 31465

In [0]:
%sql
CREATE OR REPLACE TABLE prod_data.store.store_sales_order_detail (
  SalesOrderDetailID INT COMMENT "Unique order identifier",
  SalesOrderID INT COMMENT "Order reference",
  OrderQty INT COMMENT "Amount ordered",
  ProductID INT COMMENT "Product reference",
  UnitPrice DECIMAL(19,4) COMMENT "Base price.",             --Decimal with low precision(monetary value)
  UnitPriceDiscount DECIMAL(19,4) COMMENT "Discount price.", --Decimal with low precision(monetary value)
  
  CONSTRAINT PK_store_sales_order_detail PRIMARY KEY (SalesOrderDetailID),
  CONSTRAINT FK_detail_header FOREIGN KEY (SalesOrderID)
    REFERENCES prod_data.store.store_sales_order_header (SalesOrderID),
  CONSTRAINT FK_detail_product FOREIGN KEY (ProductID)
    REFERENCES prod_data.store.store_products (ProductID)
)
USING DELTA;

In [0]:
%sql
INSERT INTO prod_data.store.store_sales_order_detail
SELECT 
  SalesOrderDetailID,
  SalesOrderID,
  OrderQty,
  ProductID,
  UnitPrice,
  UnitPriceDiscount
FROM 
  prod_data.raw.raw_sales_order_detail

num_affected_rows num_inserted_rows 121317 121317

In [0]:
%sql
DESCRIBE TABLE prod_data.store.store_products;
--DESCRIBE TABLE prod_data.store.store_sales_order_header
--DESCRIBE TABLE prod_data.store.store_sales_order_detail

col_name data_type comment ProductID int Primary key. Unique identifier for each product ProductDesc string Normal description ProductNumber string Alphanumeric column / SKU MakeFlag boolean Internal flag Color string Color of the product SafetyStockLevel int Minimum stock level required ReorderPoint int Reorder threshold StandardCost decimal(19,4) Internal product cost reference ListPrice decimal(19,4) Product price Size string Size label SizeUnitMeasureCode string Unit of measure for Size Weight decimal(10,4) Product weight. WeightUnitMeasureCode string Unit of measure for Weight ProductCategoryName string Category name of a product ProductSubCategoryName string SubCategory name of ap roduct

# - 03.Product Master Transformations:
## Perform the following transformations on the product master data and write the results into a table named publish_product:
##### 01.Replace NULL values in the Color field with N/A.
##### 02.Enhance the ProductCategoryName field when it is NULL using the following logic:
##### 02.1.If ProductSubCategoryName is in (‘Gloves’, ‘Shorts’, ‘Socks’, ‘Tights’, ‘Vests’), set ProductCategoryName to ‘Clothing’.
##### 02.2.If ProductSubCategoryName is in (‘Locks’, ‘Lights’, ‘Headsets’, ‘Helmets’, ‘Pedals’, ‘Pumps’), set ProductCategoryName to ‘Accessories’.
##### 02.3.If ProductSubCategoryName contains the word ‘Frames’ or is in (‘Wheels’, ‘Saddles’), set ProductCategoryName to ‘Components’.

In [0]:
%sql
CREATE SCHEMA IF NOT EXISTS prod_data.publish;

In [0]:
%sql
CREATE OR REPLACE TABLE prod_data.publish.publish_products (
  ProductID INT COMMENT "Primary key. Unique identifier for each product",
  ProductDesc STRING COMMENT "Normal description",
  ProductNumber STRING COMMENT "Alphanumeric column / SKU",
  MakeFlag BOOLEAN COMMENT "Internal flag",
  Color STRING COMMENT "Color of the product",
  SafetyStockLevel INT COMMENT "Minimum stock level required",
  ReorderPoint INT COMMENT "Reorder threshold",
  StandardCost DECIMAL(19,4) COMMENT "Internal product cost reference", --Changed Data Type TO DECIMAL since is monetarial value.
  ListPrice DECIMAL(19,4) COMMENT "Product price", --Changed Data Type TO DECIMAL since the representation is a fraction.
  Size STRING COMMENT "Size label",
  SizeUnitMeasureCode STRING COMMENT "Unit of measure for Size",
  Weight DECIMAL(10,4) COMMENT "Product weight.", --Changed Data Type TO DECIMAL and precision 4 since might be needed for any calculation.
  WeightUnitMeasureCode STRING COMMENT "Unit of measure for Weight",
  ProductCategoryName STRING COMMENT "Category name of a product",
  ProductSubCategoryName STRING COMMENT "SubCategory name of ap roduct", 
  CONSTRAINT PK_store_products PRIMARY KEY (ProductID)
)
USING DELTA;

In [0]:
%sql
INSERT INTO prod_data.publish.publish_products
SELECT
  ProductID,
  ProductDesc,
  ProductNumber,
  MakeFlag,
  -- 01. Replace NULL Color with 'N/A'
  COALESCE(Color, 'N/A') AS Color,
  SafetyStockLevel,
  ReorderPoint,
  StandardCost,
  ListPrice,
  Size,
  SizeUnitMeasureCode,
  Weight,
  WeightUnitMeasureCode,
  -- 02. Enhance ProductCategoryName only when it is NULL
  CASE
    WHEN ProductCategoryName IS NOT NULL 
      THEN ProductCategoryName
    WHEN ProductSubCategoryName IN ('Gloves', 'Shorts', 'Socks', 'Tights', 'Vests')
      THEN 'Clothing'
    WHEN ProductSubCategoryName IN ('Locks', 'Lights', 'Headsets', 'Helmets', 'Pedals', 'Pumps')
      THEN 'Accessories'
    WHEN ProductSubCategoryName LIKE '%Frames%'
         OR ProductSubCategoryName IN ('Wheels', 'Saddles')
      THEN 'Components'
    ELSE 'Products_with_no_description'  -- Replacing with a new category instead of just NULL
  END AS ProductCategoryName,
ProductSubCategoryName
FROM 
  prod_data.store.store_products;

num_affected_rows num_inserted_rows 295 295

In [0]:
%sql
SELECT *
FROM 
  prod_data.publish.publish_products
LIMIT 100

ProductID ProductDesc ProductNumber MakeFlag Color SafetyStockLevel ReorderPoint StandardCost ListPrice Size SizeUnitMeasureCode Weight WeightUnitMeasureCode ProductCategoryName ProductSubCategoryName 680 HL Road Frame - Black, 58 FR-R92B-58 true Black 500 375 1059.3100 1431.5000 58 CM 2.2400 LB Components Road Frames 706 HL Road Frame - Red, 58 FR-R92R-58 true Red 500 375 1059.3100 1431.5000 58 CM 2.2400 LB Components Road Frames 707 Sport-100 Helmet, Red HL-U509-R false Red 4 3 13.0863 34.9900 null null null null Accessories Helmets 708 Sport-100 Helmet, Black HL-U509 false Black 4 3 13.0863 34.9900 null null null null Accessories Helmets 709 Mountain Bike Socks, M SO-B909-M false White 4 3 3.3963 9.5000 M null null null Clothing Socks 710 Mountain Bike Socks, L SO-B909-L false White 4 3 3.3963 9.5000 L null null null Clothing Socks 711 Sport-100 Helmet, Blue HL-U509-B false Blue 4 3 13.0863 34.9900 null null null null Accessories Helmets 712 AWC Logo Cap CA-1098 false Multi 4 3 6.9223 8.9900 null null null null Products_with_no_description Caps 713 Long-Sleeve Logo Jersey, S LJ-0192-S false Multi 4 3 38.4923 49.9900 S null null null Clothing Shirt 714 Long-Sleeve Logo Jersey, M LJ-0192-M false Multi 4 3 38.4923 49.9900 M null null null Clothing Shirt 715 Long-Sleeve Logo Jersey, L LJ-0192-L false Multi 4 3 38.4923 49.9900 L null null null Clothing Shirt 716 Long-Sleeve Logo Jersey, XL LJ-0192-X false Multi 4 3 38.4923 49.9900 XL null null null Clothing Shirt 717 HL Road Frame - Red, 62 FR-R92R-62 true Red 500 375 868.6342 1431.5000 62 CM 2.3000 LB Components Road Frames 718 HL Road Frame - Red, 44 FR-R92R-44 true Red 500 375 868.6342 1431.5000 44 CM 2.1200 LB Components Road Frames 719 HL Road Frame - Red, 48 FR-R92R-48 true Red 500 375 868.6342 1431.5000 48 CM 2.1600 LB Components Road Frames 720 HL Road Frame - Red, 52 FR-R92R-52 true Red 500 375 868.6342 1431.5000 52 CM 2.2000 LB Components Road Frames 721 HL Road Frame - Red, 56 FR-R92R-56 true Red 500 375 868.6342 1431.5000 56 CM 2.2400 LB Components Road Frames 722 LL Road Frame - Black, 58 FR-R38B-58 true Black 500 375 204.6251 337.2200 58 CM 2.4600 LB Components Road Frames 723 LL Road Frame - Black, 60 FR-R38B-60 true Black 500 375 204.6251 337.2200 60 CM 2.4800 LB Components Road Frames 724 LL Road Frame - Black, 62 FR-R38B-62 true Black 500 375 204.6251 337.2200 62 CM 2.5000 LB Components Road Frames 725 LL Road Frame - Red, 44 FR-R38R-44 true Red 500 375 187.1571 337.2200 44 CM 2.3200 LB Components Road Frames 726 LL Road Frame - Red, 48 FR-R38R-48 true Red 500 375 187.1571 337.2200 48 CM 2.3600 LB Components Road Frames 727 LL Road Frame - Red, 52 FR-R38R-52 true Red 500 375 187.1571 337.2200 52 CM 2.4000 LB Components Road Frames 728 LL Road Frame - Red, 58 FR-R38R-58 true Red 500 375 187.1571 337.2200 58 CM 2.4600 LB Components Road Frames 729 LL Road Frame - Red, 60 FR-R38R-60 true Red 500 375 187.1571 337.2200 60 CM 2.4800 LB Components Road Frames 730 LL Road Frame - Red, 62 FR-R38R-62 true Red 500 375 187.1571 337.2200 62 CM 2.5000 LB Components Road Frames 731 ML Road Frame - Red, 44 FR-R72R-44 true Red 500 375 352.1394 594.8300 44 CM 2.2200 LB Components Road Frames 732 ML Road Frame - Red, 48 FR-R72R-48 true Red 500 375 352.1394 594.8300 48 CM 2.2600 LB Components Road Frames 733 ML Road Frame - Red, 52 FR-R72R-52 true Red 500 375 352.1394 594.8300 52 CM 2.3000 LB Components Road Frames 734 ML Road Frame - Red, 58 FR-R72R-58 true Red 500 375 352.1394 594.8300 58 CM 2.3600 LB Components Road Frames 735 ML Road Frame - Red, 60 FR-R72R-60 true Red 500 375 352.1394 594.8300 60 CM 2.3800 LB Components Road Frames 736 LL Road Frame - Black, 44 FR-R38B-44 true Black 500 375 204.6251 337.2200 44 CM 2.3200 LB Components Road Frames 737 LL Road Frame - Black, 48 FR-R38B-48 true Black 500 375 204.6251 337.2200 48 CM 2.3600 LB Components Road Frames 738 LL Road Frame - Black, 52 FR-R38B-52 true Black 500 375 204.6251 337.2200 52 CM 2.4000 LB Components Road Frames

# - 04.Sales Order Transformations:
## Join SalesOrderDetail with SalesOrderHeader on SalesOrderId and apply the following transformations:
##### 01.Calculate LeadTimeInBusinessDays as the difference between OrderDate and ShipDate, excluding Saturdays and Sundays.
##### 02.Calculate TotalLineExtendedPrice using the formula: OrderQty * (UnitPrice - UnitPriceDiscount).
##### 03.Write the results into a table named publish_orders, including:
##### 04.All fields from SalesOrderDetail.
##### 05.All fields from SalesOrderHeader except SalesOrderId, and rename Freight to TotalOrderFreight.

In [0]:
%sql
--store.Tables: SalesOrdeDetails X SalesOrderHeader
CREATE OR REPLACE TABLE prod_data.publish.publish_orders (
  -- From SalesOrderDetail
  SalesOrderDetailID INT COMMENT "Primary key. Unique identifier for each order line.",
  SalesOrderID INT COMMENT "FK reference to the sales order header.",
  OrderQty INT COMMENT "Quantity ordered.",
  ProductID INT COMMENT "FK reference to the product.",
  UnitPrice DECIMAL(19,4) COMMENT "Base unit price before discount.",
  UnitPriceDiscount DECIMAL(19,4) COMMENT "Discount applied on the unit price.",
  TotalLineExtendedPrice DECIMAL(19,4) COMMENT "Calculated:OrderQty * UnitPrice * (1 - UnitPriceDiscount).",

  -- From SalesOrderHeader
  OrderDate DATE COMMENT "Date the order was placed.",
  ShipDate DATE COMMENT "Date the order was shipped.",
  OnlineOrderFlag BOOLEAN COMMENT "Flag indicating if order was placed online.",
  AccountNumber STRING COMMENT "ERP customer account identifier.",
  CustomerID INT COMMENT "Customer identifier.",
  SalesPersonID INT COMMENT "Salesperson identifier.",
  TotalOrderFreight DECIMAL(19,4) COMMENT "Freight cost at order level.",

  -- Calculated(From requirement)
  LeadTimeInBusinessDays INT COMMENT "Business-day difference between OrderDate and ShipDate (excludes weekends).",
  CONSTRAINT pk_publish_orders PRIMARY KEY (SalesOrderDetailID)
)
USING DELTA;


In [0]:
%sql
INSERT INTO prod_data.publish.publish_orders
SELECT
  d.SalesOrderDetailID,
  d.SalesOrderID,
  d.OrderQty,
  d.ProductID,
  d.UnitPrice,
  d.UnitPriceDiscount,
  -- 02. TotalLineExtendedPrice = OrderQty * (UnitPrice - UnitPriceDiscount)

  CAST(d.OrderQty * d.UnitPrice * (1 - d.UnitPriceDiscount) AS DECIMAL(19,4)) AS TotalLineExtendedPrice,

  h.OrderDate,
  h.ShipDate,
  h.OnlineOrderFlag,
  h.AccountNumber,
  h.CustomerID,
  h.SalesPersonID,
  CAST(h.Freight AS DECIMAL(19,4)) AS TotalOrderFreight,

  -- 01. LeadTimeInBusinessDays (diferença em dias úteis entre OrderDate e ShipDate)
  SIZE(
    FILTER(
      SEQUENCE(h.OrderDate, h.ShipDate),
      d -> date_format(d, 'E') NOT IN ('Sat', 'Sun')
    )
  ) AS LeadTimeInBusinessDays
FROM 
  prod_data.store.store_sales_order_detail d
INNER JOIN prod_data.store.store_sales_order_header h ON d.SalesOrderID = h.SalesOrderID;

num_affected_rows num_inserted_rows 121317 121317

In [0]:
%sql
SELECT *
FROM 
  prod_data.publish.publish_orders
LIMIT 100

SalesOrderDetailID SalesOrderID OrderQty ProductID UnitPrice UnitPriceDiscount TotalLineExtendedPrice OrderDate ShipDate OnlineOrderFlag AccountNumber CustomerID SalesPersonID TotalOrderFreight LeadTimeInBusinessDays 1 43659 1 776 2024.9940 0.0000 2024.9940 2021-05-31 2021-06-07 false 10-4020-000676 29825 279 616.0984 6 2 43659 3 777 2024.9940 0.0000 6074.9820 2021-05-31 2021-06-07 false 10-4020-000676 29825 279 616.0984 6 3 43659 1 778 2024.9940 0.0000 2024.9940 2021-05-31 2021-06-07 false 10-4020-000676 29825 279 616.0984 6 4 43659 1 771 2039.9940 0.0000 2039.9940 2021-05-31 2021-06-07 false 10-4020-000676 29825 279 616.0984 6 5 43659 1 772 2039.9940 0.0000 2039.9940 2021-05-31 2021-06-07 false 10-4020-000676 29825 279 616.0984 6 6 43659 2 773 2039.9940 0.0000 4079.9880 2021-05-31 2021-06-07 false 10-4020-000676 29825 279 616.0984 6 7 43659 1 774 2039.9940 0.0000 2039.9940 2021-05-31 2021-06-07 false 10-4020-000676 29825 279 616.0984 6 8 43659 3 714 28.8404 0.0000 86.5212 2021-05-31 2021-06-07 false 10-4020-000676 29825 279 616.0984 6 9 43659 1 716 28.8404 0.0000 28.8404 2021-05-31 2021-06-07 false 10-4020-000676 29825 279 616.0984 6 10 43659 6 709 5.7000 0.0000 34.2000 2021-05-31 2021-06-07 false 10-4020-000676 29825 279 616.0984 6 11 43659 2 712 5.1865 0.0000 10.3730 2021-05-31 2021-06-07 false 10-4020-000676 29825 279 616.0984 6 12 43659 4 711 20.1865 0.0000 80.7460 2021-05-31 2021-06-07 false 10-4020-000676 29825 279 616.0984 6 13 43660 1 762 419.4589 0.0000 419.4589 2021-05-31 2021-06-07 false 10-4020-000117 29672 279 38.8276 6 14 43660 1 758 874.7940 0.0000 874.7940 2021-05-31 2021-06-07 false 10-4020-000117 29672 279 38.8276 6 15 43661 1 745 809.7600 0.0000 809.7600 2021-05-31 2021-06-07 false 10-4020-000442 29734 282 985.5530 6 16 43661 1 743 714.7043 0.0000 714.7043 2021-05-31 2021-06-07 false 10-4020-000442 29734 282 985.5530 6 17 43661 2 747 714.7043 0.0000 1429.4086 2021-05-31 2021-06-07 false 10-4020-000442 29734 282 985.5530 6 18 43661 4 712 5.1865 0.0000 20.7460 2021-05-31 2021-06-07 false 10-4020-000442 29734 282 985.5530 6 19 43661 4 715 28.8404 0.0000 115.3616 2021-05-31 2021-06-07 false 10-4020-000442 29734 282 985.5530 6 20 43661 2 742 722.5949 0.0000 1445.1898 2021-05-31 2021-06-07 false 10-4020-000442 29734 282 985.5530 6 21 43661 3 775 2024.9940 0.0000 6074.9820 2021-05-31 2021-06-07 false 10-4020-000442 29734 282 985.5530 6 22 43661 2 778 2024.9940 0.0000 4049.9880 2021-05-31 2021-06-07 false 10-4020-000442 29734 282 985.5530 6 23 43661 2 711 20.1865 0.0000 40.3730 2021-05-31 2021-06-07 false 10-4020-000442 29734 282 985.5530 6 24 43661 2 741 0.0000 0.0000 0.0000 2021-05-31 2021-06-07 false 10-4020-000442 29734 282 985.5530 6 25 43661 4 776 2024.9940 0.0000 8099.9760 2021-05-31 2021-06-07 false 10-4020-000442 29734 282 985.5530 6 26 43661 2 773 2039.9940 0.0000 4079.9880 2021-05-31 2021-06-07 false 10-4020-000442 29734 282 985.5530 6 27 43661 2 716 28.8404 0.0000 57.6808 2021-05-31 2021-06-07 false 10-4020-000442 29734 282 985.5530 6 28 43661 2 777 2024.9940 0.0000 4049.9880 2021-05-31 2021-06-07 false 10-4020-000442 29734 282 985.5530 6 29 43661 5 708 20.1865 0.0000 100.9325 2021-05-31 2021-06-07 false 10-4020-000442 29734 282 985.5530 6 30 43662 3 764 419.4589 0.0000 1258.3767 2021-05-31 2021-06-07 false 10-4020-000227 29994 282 867.2389 6 31 43662 5 770 419.4589 0.0000 2097.2945 2021-05-31 2021-06-07 false 10-4020-000227 29994 282 867.2389 6 32 43662 2 730 183.9382 0.0000 367.8764 2021-05-31 2021-06-07 false 10-4020-000227 29994 282 867.2389 6 33 43662 4 754 874.7940 0.0000 3499.1760 2021-05-31 2021-06-07 false 10-4020-000227 29994 282 867.2389 6 34 43662 3 725 183.9382 0.0000 551.8146 2021-05-31 2021-06-07 false 10-4020-000227 29994 282 867.2389 6 35 43662 5 762 419.4589 0.0000 2097.2945 2021-05-31 2021-06-07 false 10-4020-000227 29994 282 867.2389 6 36 43662 3 765 419.4589 0.0000 1258.3767 2021-05-31 2021-06-07 false 10-4020-000227 29994 282 867.2389 6 37 43662 2 768 419.4589 0.0000 838.9178 2021

# - 05.Data Analysis Questions:
## Provide answers to the following questions based on the transformed data:
##### 01.Which color generated the highest revenue each year?
##### 02.What is the average LeadTimeInBusinessDays by ProductCategoryName?

In [0]:
%sql
-- 01. Highest revenue color by year
WITH color_year_revenue AS (
  SELECT
    YEAR(o.OrderDate) AS order_year,
    p.Color,
    SUM(o.TotalLineExtendedPrice) AS total_revenue
  FROM 
    prod_data.publish.publish_orders o
    INNER JOIN prod_data.publish.publish_products p ON o.ProductID = p.ProductID
  GROUP BY 
    YEAR(o.OrderDate), p.Color
),
rank_highest_revenue_year AS (
  SELECT
    order_year,
    Color,
    total_revenue,
    ROW_NUMBER() OVER (PARTITION BY order_year ORDER BY total_revenue DESC) AS rk
  FROM 
    color_year_revenue
)
SELECT
  order_year,
  Color,
  total_revenue
FROM 
  rank_highest_revenue_year
WHERE 
  rk = 1
ORDER BY 
  total_revenue DESC, order_year DESC;

order_year Color total_revenue 2023 Black 15031199.7343 2022 Black 13917844.9058 2024 Yellow 6360243.0023 2021 Red 6019394.4108

In [0]:
%sql
-- 02. Average lead time (business days) by product category
SELECT
  p.ProductCategoryName,
  AVG(o.LeadTimeInBusinessDays) AS avg_LeadTimeInBusinessDays
FROM 
  prod_data.publish.publish_orders o
  INNER JOIN prod_data.publish.publish_products p ON o.ProductID = p.ProductID
GROUP BY 
  p.ProductCategoryName
ORDER BY 
  avg_LeadTimeInBusinessDays DESC;


ProductCategoryName avg_LeadTimeInBusinessDays Products_with_no_description 5.7202082171407325 Clothing 5.709380234505863 Accessories 5.702787804316105 Bikes 5.667897567632656 Components 5.667113624438874